---
title: "class18"
author: "Darren Lam"
date: "11/26/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Using GenomicDataCommons, TCGAbiolinks, and maftools to explore public cancer genomic data

## Installing and loading packages
```{r}
# BiocManager::install( c("GenomicDataCommons", "TCGAbiolinks", "maftools"))
# library(GenomicDataCommons)
# library(TCGAbiolinks)
# library(maftools)
```

Checking on GDC status
```{r}
# status() # expect to see $status [1] "OK"
```

## Querying  GDC metadata from R

Querying projects metadata
```{r}
projects <- getGDCprojects()
head(projects)
```

Querying cases metadata
```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations
head(cases_by_project)
```

Making a barplot of the cases per project
```{r}
x <- cases_by_project$project.project_id
head(x)

colvec <- rep("lightblue", nrow(x))
colvec[x$key == "TCGA-PAAD"] <- rep("red")

barplot(x$doc_count, names.arg=x$key, log="y", col=colvec, las=2)
```

Summarizing a project
```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")
head(samp)
```

# Comparing healthy and tumor tissue protein sequences

Reading in FASTA files
```{r}
library(bio3d)
seqs <- read.fasta("lecture18_sequences.fa")
```

```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites

## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names

## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```

